app/grandchallenge/evaluation/models.py

```diff
@@ -858,6 +895,59 @@ def send_give_algorithm_editors_job_view_permissions_changed_email(self):
             subscription_type=EmailSubscriptionTypes.SYSTEM,
         )
 
+    @cached_property
+    def descendants(self):
+        descendants = []
+        children = self.children.all()
+        for child in children:
+            descendants.append(child)
+            descendants.extend(child.descendants)
+        return descendants
+
+    @cached_property
+    def parent_phase_choices(self):
+        extra_filters = {}
+        extra_annotations = {}
+        if self.submission_kind == SubmissionKindChoices.ALGORITHM:
+            algorithm_inputs = self.algorithm_inputs.all()
+            algorithm_outputs = self.algorithm_outputs.all()
+            extra_annotations = {
+                "total_input_count": Count("algorithm_inputs", distinct=True),
+                "total_output_count": Count(
+                    "algorithm_outputs", distinct=True
+                ),
+                "relevant_input_count": Count(
+                    "algorithm_inputs",
+                    filter=Q(algorithm_inputs__in=algorithm_inputs),
+                    distinct=True,
+                ),
+                "relevant_output_count": Count(
+                    "algorithm_outputs",
+                    filter=Q(algorithm_outputs__in=algorithm_outputs),
+                    distinct=True,
+                ),
+            }
+            extra_filters = {
+                "total_input_count": len(algorithm_inputs),
+                "total_output_count": len(algorithm_outputs),
+                "relevant_input_count": len(algorithm_inputs),
+                "relevant_output_count": len(algorithm_outputs),
+            }
```

Comment:
I think I recognise this code from somewhere else? Or am I thinking of algorithm jobs and archive items? Are there any possibilities for reuse?

Reply:
Yes, this is very similar to what we do in `user_algorithms_for_phase`. The lookups are different (`inputs` vs. `algorithm_inputs` etc), so we couldn't directly reuse, but I could write a method to return the annotation and filter dictionaries based on the lookup parameters.

Reply:
Okay! Then no, no need to reuse.